hypothesis/lms
# Add `DigestService`

Add a new `DigestService` for generating and sending instructor email digests and change the celery task to call this service.

This gets us to the point where the email digests admin page can be used to really send an email through Mailchimp.

After this it only remains to write the periodic task that will send emails to users every night.

### This is egregiously inefficient with DB queries

The code does a separate query for each `h_userid` (to find all `User`'s with that `h_userid`), two separate queries for each `authority_provided_id` (to find all the `Course`'s with that `authority_provided_id` or with the `authority_provided_id` of the parent grouping in the case of sub-groupings), and a separate query for each `(user, course)` combination to ask whether that user is an instructor in that course. The results of these queries aren't cached either so it may even do the same query again and again (for example when the same user has many annotations in the requested timeframe).

Hopefully this is good enough at least for us to test the feature using the admin page so that I can get the remaining bits of this feature and the Celery and Mailchimp setups done etc. Maybe it's even good enough to beta test the feature with a small cohort of real users.

But I think we are going to have to follow up this PR with some refactoring to make the number of DB queries scale better with respect to the number of annotations. In Jon's gist the idea was that you do one query to get all the users at once, another to get all the courses at once, and another to get a mapping of all the courses to all their instructors all at once, and then the business logic code just refers to these Python dicts whenever it needs to get a user or course or ask whether a user is an instructor in its course. The code will need to be refactored a bit to enable this fixed-number-of-queries approach as the current code organization isn't amenable to it but I think that refactoring will hopefully be fairly easy and I have some ideas for how I want to do it. What I don't know yet is how to actually write these all-at-once versions of the DB queries.

Testing
=======

* You'll the latest version of h's `main` branch running because you need the new bulk annotation API
* You might need to log in to <https://hypothesis.instructure.com/> as `eng+canvasteacher@hypothes.is` and launch [a localhost assignment](https://hypothesis.instructure.com/courses/125/assignments/873) in order to create the instructor record in your local DB

Nothing should happen if there aren't any annotations
-----------------------------------------------------

1. Go to <http://localhost:8001/admin/email>, enter your email address, and submit the form (the default values will do).

2. Nothing should be printed in the terminal: it shouldn't send any emails if there aren't any annotations.

Nothing should happen if there are only instructor annotations
--------------------------------------------------------------

1. Log in to <https://hypothesis.instructure.com/> as the `eng+canvasteacher@hypothes.is` account
2. Launch [a localhost assignment](https://hypothesis.instructure.com/courses/125/assignments/873)
3. Create some annotations
4. Submit the admin form again. Choose since and until times that cover the time at which you created the annotations. For example you can just yesterday and tomorrow.
5. Nothing should be printed in the terminal: it shouldn't send any emails if there are only instructor annotations

It should send an email if there are student annotations
--------------------------------------------------------

1. Log in to <https://hypothesis.instructure.com/> as the `eng+canvasstudent@hypothes.is` account
2. Launch [a localhost assignment](https://hypothesis.instructure.com/courses/125/assignments/873)
3. Create some annotations
4. Submit the admin form again. Again, choose since and until times that cover the time at which you created the annotations.
5. It won't actually send any emails since you don't have a `MAILCHIMP_API_KEY` configured locally but you should see `MailchimpService` logging from the worker process what it would have sent to Mailchimp:

   ```
   mailchimp_client.send_template(...)
   ```
